src/bt_peer_connection.cpp

```diff
@@ -1335,7 +1335,7 @@ namespace libtorrent
 		span<char const> recv_buffer = m_recv_buffer.get();
 
 		const char* ptr = recv_buffer.begin() + 1;
-		piece_index_t const piece(detail::read_uint32(ptr));
+		piece_index_t const piece(aux::numeric_cast<int>(detail::read_uint32(ptr)));
```

Comment:
I'm pretty sure there's a run-time check in ``incoming_suggest()`` to validate the piece index. the test that's failing, most likely sends a piece index > ``INT_MAX``. In this case, it's first read as unsigned, then cast to signed, then checked. your ``numeric_cast<>`` here introduces an assert in front of the run-time validation, and hence the test for it will trigger your assert instead of exercising the validation.